Fix generating file names for edited files, and use wp_unique_filename()

azaozz opened the pull request:
Follow-up to https://github.com/WordPress/gutenberg/pull/23284.

## Description
Fixes few errors and inconsistencies when editing an already-edited image file, and when creating the new attachment post. Also adds some better errors and some cleanup of vars, names, etc. and still has few TBD/TODO.

## How has this been tested?
When testing please monitor what files are created in the uploads dir (`wp-content/uploads/year/month/`).

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

azaozz committed:
Fix generating file names fro edited files, and use wp_unique_filename()

Also some better errors and some cleanup of names.

ellatrix commented on lib/class-wp-rest-image-editor-controller.php:
```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
+			// TODO: return the "real" error message?
+			if ( is_wp_error( $result ) ) {
+				return new WP_Error( 'rest_image_rotation_failed', __( 'Unable to rotate this image.', 'gutenberg' ), array( 'status' => 500 ) );
+			}
```
What if it doesn't load and there is no error?

ellatrix commented on lib/class-wp-rest-image-editor-controller.php:
```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
+		// The user may be editing a previously edited image.
+		if ( preg_match( '/-edited(-\d+)?$/', $image_name ) ) {
+			// Remove any `-1`, `-2`, etc. `wp_unique_filename()` will add the proper number.
+			$image_name = preg_replace( '/-edited(-\d+)?$/', '-edited', $image_name );
```
Can't we replace immediately? Do we need to match first?

azaozz replied:
Need it for the `else` to append`-edited`.

ellatrix commented on lib/class-wp-rest-image-editor-controller.php:
```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
 		}
 
-		// Update attachment details.
+		// Create new attachment post.
```
Should we copy the rest of the attachment data?

azaozz replied:
Yeah, may make sense. But then we'll get two (or more) attachments with the same description and caption. In general don't think the image alt text should be carried over as alt has to be contextual. Editing an older, already used image would suggest it is being used in different context. Editing an image just after uploading it won't have alt text.

ellatrix commented on lib/class-wp-rest-image-editor-controller.php:
```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
-			return new WP_Error( 'rest_cannot_find_attached_file', __( 'Unable to find original media file.', 'gutenberg' ), array( 'status' => 500 ) );
-		}
+		if ( isset( $params['rotation'] ) ) { // TODO: use `rotate` instead?
+			// CW / CCW
```
Might be good to write it in full :)

ellatrix commented on lib/class-wp-rest-image-editor-controller.php:
```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
 
-		$media_id = $params['media_id'];
+		// This also confirms the attachment is an image.
+		$image_file = wp_get_original_image_path( $attachment_id );
```
Might be good to leave a comment saying that this is the original uploaded file.

ellatrix commented on lib/class-wp-rest-image-editor-controller.php:
```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
+		// Create the uploads sub-directory if needed.
+		$uploads = wp_upload_dir();
 
-		$filename = rtrim( dirname( $image_path ), '/' ) . '/' . $target_file;
```
Previously we were overriding the last edited file, right?

azaozz replied:
Yeah, kind of. Was pretty messy. Overwriting the last edited file if it is in the same uploads subdirectory, but not all sub-sizes of it. If the edited file was smaller, and not all sub-sizes were created, some of the old sub-sizes were left "orphaned".

Better workflow and UX would be to keep all edited files, perhaps let the user know they can delete unused files from the media library.

ellatrix reviewed (approved):
Looks good to me! Just left some questions.

azaozz commented on lib/class-wp-rest-image-editor-controller.php:
```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
+			$result = $image_editor->rotate( 0 - $params['rotation'] );
 
-		$image_editor = wp_get_image_editor( $image_path );
-		if ( ! $image_editor->load() ) {
```
No need. `wp_get_image_editor()` instantiates the (proper) editor and loads the file passed as first arg. It always returns a WP_Error on failure. See https://developer.wordpress.org/reference/functions/wp_get_image_editor/.

azaozz committed:
Fix cropping when very small changes, show error when saving image that has not been edited.

azaozz committed:
Fix cropping.

- Move the check if the crop dimensions are > 0.1% to the js.
- Test if the crop dimensions are whitin bounds in PHP.
- Make crop values optional.

azaozz committed:
Fix when to crop test.

No need to look at `x` and `y`. If they have changed, the width or/and height have changed too.
Also fix white space in (ugly) array alignment...

TimothyBJacobs commented on lib/class-wp-rest-image-editor-controller.php:
```diff
@@ -101,90 +98,157 @@ public function permission_callback( $request ) {
+				( $params['x'] >= 0 && $params['x'] < 100 ) &&
+				( $params['y'] >= 0 && $params['y'] < 100 ) &&
+				( $params['width'] >= 1 && $params['width'] <= 100 ) &&
+				( $params['height'] >= 1 && $params['height'] <= 100 )
```
We should put these bounds checks in the parameter args definition. We can use `maximum` to set a max value. You may need to use `exclusiveMaximum` for the x and y.

azaozz replied:
Sounds good. Was just a bit unsure if floats would work, but with `width` and `height` expecting minimum `1`, `x` and `y` can be set to expect maximum `99` (inclusive). (Cannot have an image cropped to 1% width but starting point be over 99% of the width, etc.)

azaozz replied:
Thinking a bit more, the API should only care if these numbers are percents, 0 - 100. The rest should be handled when doing the cropping.

azaozz committed:
Add 0-100 min/max for the crop values

TimothyBJacobs commented on lib/class-wp-rest-image-editor-controller.php:
```diff
@@ -101,90 +104,149 @@ public function permission_callback( $request ) {
-		if ( isset( $attachment_info['richimage'] ) ) {
-			$meta = array_merge( $meta, $attachment_info['richimage'] );
+		if ( ! $rotate && ! $crop ) {
+			$error = __( 'The image was not erdited. Edit the image before applying the changes.', 'gutenberg' );
```
```suggestion
			$error = __( 'The image was not edited. Edit the image before applying the changes.', 'gutenberg' );
```

azaozz replied:
Ouch, thanks! Would a 400 be appropriate error there? At this point it's an "empty" request, nothing to do.

TimothyBJacobs replied:
Yeah I think a `400` makes sense.

azaozz committed:
Fix typo.

Co-authored-by: Timothy Jacobs <timothy@ironbounddesigns.com>
